src/realm/util/ez_websocket.hpp

```diff
@@ -0,0 +1,115 @@
+#ifndef REALM_UTIL_EZ_WEBSOCKET_HPP
+#define REALM_UTIL_EZ_WEBSOCKET_HPP
+
+#include <random>
+#include <system_error>
+#include <map>
+
+#include <realm/sync/config.hpp>
+#include <realm/util/http.hpp>
+
+namespace realm::util::network {
+class Service;
+}
+
+
+namespace realm::util::websocket {
+using port_type = sync::port_type;
+
+// TODO figure out what belongs on config and what belongs on endpoint.
+struct EZConfig {
+    util::Logger& logger;
+    std::mt19937_64& random;
+    util::network::Service& service;
+    std::string user_agent;
+    bool tcp_no_delay;
+};
+
+struct EZEndpoint {
+    std::string address;
+    port_type port;
+    std::string http_host; // Contents of `Host:` request header
+    std::string path;
+    std::string protocols; // separated with ", "
+    bool is_ssl;
+    util::HTTPHeaders headers; // Includes both auth and "custom" headers
+    bool verify_servers_ssl_certificate;
+    util::Optional<std::string> ssl_trust_certificate_path;
+    std::function<SyncConfig::SSLVerifyCallback> ssl_verify_callback;
+    util::Optional<SyncConfig::ProxyConfig> proxy;
+};
+
+class EZObserver {
```

Comment:
So is this excluding the interface that SDKs will have to conform to? Trying to forward think, but this seems odd:

```cpp
public CocoaWebSocket: EZSocket, EZObserver {
    /// methods that write and methods that receive events
}
```

Reply:
No, a `CocoaWebSocket` would only implement the (to be renamed) `EZSocket` interface. The `EZObserver` interface is implemented by the consumer of the socket. It is what the socket implementation sends events to.